packages/utils/src/eslint-utils/rule-tester/RuleTester.ts

```diff
@@ -0,0 +1,259 @@
+      testOptions,
+    ) as TSESLint.ParserOptions;
+    const filename = `file.ts${resolvedOptions.ecmaFeatures?.jsx ? 'x' : ''}`;
+    if (resolvedOptions.project) {
+      return path.join(
+        resolvedOptions.tsconfigRootDir != null
+          ? resolvedOptions.tsconfigRootDir
+          : process.cwd(),
+        filename,
+      );
+    }
+    return filename;
+  }
+
+  // as of eslint 6 you have to provide an absolute path to the parser
+  // If you don't do that at the test level, the test will fail somewhat cryptically...
+  // This is a lot more explicit
+  run<TMessageIds extends string, TOptions extends Readonly<unknown[]>>(
+    name: string,
+    rule: TSESLint.RuleModule<TMessageIds, TOptions>,
+    testsReadonly: RunTests<TMessageIds, TOptions>,
+  ): void {
+    const errorMessage = `Do not set the parser at the test level unless you want to use a parser other than ${TS_ESLINT_PARSER}`;
+
+    const tests = {
+      // standardize the valid tests as objects
+      valid: testsReadonly.valid.map(test => {
+        if (typeof test === 'string') {
+          return {
+            code: test,
+          };
+        }
+        return test;
+      }),
+      invalid: testsReadonly.invalid,
+    };
+
+    /*
+    Automatically add a filename to the tests to enable type-aware tests to "just work".
+    This saves users having to verbosely and manually add the filename to every
+    single test case.
+    Hugely helps with the string-based valid test cases as it means they don't
+    need to be made objects!
+    */
+    const addFilename = <
+      T extends
+        | ValidTestCase<TOptions>
+        | InvalidTestCase<TMessageIds, TOptions>,
+    >(
+      test: T,
+    ): T => {
+      if (test.parser === TS_ESLINT_PARSER) {
+        throw new Error(errorMessage);
+      }
+      if (!test.filename) {
+        return {
+          ...test,
+          filename: this.getFilename(test.parserOptions),
+        };
+      }
+      return test;
+    };
+    tests.valid = tests.valid.map(addFilename);
+    tests.invalid = tests.invalid.map(addFilename);
+
+    const hasOnly = ((): boolean => {
+      for (const test of tests.valid) {
+        if (test.only) {
+          return true;
+        }
+      }
+      for (const test of tests.invalid) {
+        if (test.only) {
+          return true;
+        }
+      }
+      return false;
+    })();
+    // if there is an `only: true` - don't apply constraints - assume that
+    // we are in "local development" mode rather than "CI validation" mode
+    if (!hasOnly) {
+      /*
+      Automatically skip tests that don't satisfy the dependency constraints.
+      */
+      const hasConstraints = ((): boolean => {
+        for (const test of tests.valid) {
+          if (
+            test.dependencyConstraints &&
+            Object.keys(test.dependencyConstraints).length > 0
+          ) {
+            return true;
+          }
+        }
+        for (const test of tests.invalid) {
```

Comment:
I was purposely avoiding spreading because spreading would mean iterating all tests once before you can iterate again to do your checks.

I felt like gaining perf in exchange for one copied loop was worth the trade-off.

Thinking about it now - one way we could do this without the spread would be to define a custom iterator object. Maybe I'll do that!

Reply:
That is a snazzy solution. Though, how many tests are we defining on average per file, and on average how many files? Is this really going to be a noticeable perf change?

Reply:
Some rule trsts do have thousands of tests in them! At the very least a few hundred would be the average.